cpp/src/arrow/compute/exec/hash_join_benchmark.cc

```diff
@@ -0,0 +1,406 @@
+std::vector<std::vector<int64_t>> selectivity_args = {
+    benchmark::CreateDenseRange(0, 100, 20), benchmark::CreateRange(1, 4 * 1024, 8)};
+
+BENCHMARK_CAPTURE(BM_HashJoinBasic_Selectivity, "{int32}", {int32()})
+    ->ArgNames(selectivity_argnames)
+    ->ArgsProduct(selectivity_args);
+BENCHMARK_CAPTURE(BM_HashJoinBasic_Selectivity, "{int64}", {int64()})
+    ->ArgNames(selectivity_argnames)
+    ->ArgsProduct(selectivity_args);
+
+// Joins on UTF8 are currently really slow, so anything above 512 doesn't finished within
+// a reasonable amount of time.
+BENCHMARK_CAPTURE(BM_HashJoinBasic_Selectivity, "{utf8}", {utf8()})
+    ->ArgNames(selectivity_argnames)
+    ->ArgsProduct({benchmark::CreateDenseRange(0, 100, 20),
+                   benchmark::CreateRange(1, 64, 8)});
+
+BENCHMARK_CAPTURE(BM_HashJoinBasic_Selectivity, "{fixed_size_binary(16)}",
+                  {fixed_size_binary(16)})
+    ->ArgNames(selectivity_argnames)
+    ->ArgsProduct(selectivity_args);
+BENCHMARK_CAPTURE(BM_HashJoinBasic_Selectivity, "{fixed_size_binary(32)}",
+                  {fixed_size_binary(32)})
+    ->ArgNames(selectivity_argnames)
+    ->ArgsProduct(selectivity_args);
+
+std::vector<std::string> jointype_argnames = {"Selectivity", "HashTable krows"};
+std::vector<std::vector<int64_t>> jointype_args = {
+    benchmark::CreateDenseRange(0, 100, 20), benchmark::CreateRange(1, 4 * 1024, 8)};
+
+BENCHMARK_CAPTURE(BM_HashJoinBasic_JoinType, "Inner", JoinType::INNER)
+    ->ArgNames(jointype_argnames)
+    ->ArgsProduct(jointype_args);
+BENCHMARK_CAPTURE(BM_HashJoinBasic_JoinType, "Left Semi", JoinType::LEFT_SEMI)
+    ->ArgNames(jointype_argnames)
+    ->ArgsProduct(jointype_args);
+BENCHMARK_CAPTURE(BM_HashJoinBasic_JoinType, "Right Semi", JoinType::RIGHT_SEMI)
+    ->ArgNames(jointype_argnames)
+    ->ArgsProduct(jointype_args);
+BENCHMARK_CAPTURE(BM_HashJoinBasic_JoinType, "Left Anti", JoinType::LEFT_ANTI)
+    ->ArgNames(jointype_argnames)
+    ->ArgsProduct(jointype_args);
+BENCHMARK_CAPTURE(BM_HashJoinBasic_JoinType, "Right Anti", JoinType::RIGHT_ANTI)
+    ->ArgNames(jointype_argnames)
+    ->ArgsProduct(jointype_args);
+BENCHMARK_CAPTURE(BM_HashJoinBasic_JoinType, "Left Outer", JoinType::LEFT_OUTER)
+    ->ArgNames(jointype_argnames)
+    ->ArgsProduct(jointype_args);
+BENCHMARK_CAPTURE(BM_HashJoinBasic_JoinType, "Right Outer", JoinType::RIGHT_OUTER)
+    ->ArgNames(jointype_argnames)
+    ->ArgsProduct(jointype_args);
+BENCHMARK_CAPTURE(BM_HashJoinBasic_JoinType, "Full Outer", JoinType::FULL_OUTER)
+    ->ArgNames(jointype_argnames)
+    ->ArgsProduct(jointype_args);
+
+BENCHMARK(BM_HashJoinBasic_MatchesPerRow)
+    ->ArgNames({"MatchesPerRow", "HashTable krows"})
+    ->ArgsProduct({benchmark::CreateRange(1, 16, 2),
+                   benchmark::CreateRange(1, 4 * 1024, 8)});
+
+BENCHMARK(BM_HashJoinBasic_PayloadSize)
+    ->ArgNames({"Payload Size", "MatchesPerRow", "HashTable krows"})
```

Comment:
Nit, but can you ensure a consistent naming convention? For example "Payload Size", "Matches Per Row", "Hash Table Cardinality".